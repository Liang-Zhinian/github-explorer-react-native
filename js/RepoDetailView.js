/**
 * MIT License
 *
 * Copyright (c) 2017 johnwakley
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

// @flow

import React, {Component} from 'react'
import {
  StyleSheet,
  Text,
  View,
  ScrollView,
  Button,
  Image,
  Linking
} from 'react-native'
import colors from './common/colors'
import {HeroText} from './components/Text'
import Badge from './components/Badge'

export default class RepoDetailView extends Component {
  static navigationOptions = {}

  props: {
    navigation: any
  }

  openRepoUrl (url: string) {
    Linking.openURL(url).catch(err => console.error('An error occurred', err))
  }

  render () {
    const {repo} = this.props.navigation.state.params
    const avatarUrl = repo.owner.avatarUrl
    const repoLang = repo.primaryLanguage

    return (
      <View style={styles.container}>
        <View style={styles.hero}>
          <HeroText>{repo.name}</HeroText>
          {Boolean(avatarUrl) &&
            <Image style={styles.avatar} source={{uri: avatarUrl}} />}
          <Text>by {repo.owner.login}</Text>
          <View style={styles.badgeRow}>
            {Boolean(repoLang) &&
              <Badge
                iconName='fiber-manual-record'
                iconColor={repoLang.color}
                label={repoLang.name}
              />}
            <Badge
              iconName='star'
              iconColor='grey'
              label={repo.stargazers.totalCount}
            />
            <Badge
              iconName='remove-red-eye'
              iconColor='grey'
              label={repo.watchers.totalCount}
            />
          </View>
        </View>
        <View style={styles.body}>
          <ScrollView style={styles.scrollView} testID='scrollview' accessibilityLabel='scrollview'>
            <Text style={{marginBottom: 20}}>{repo.description}</Text>
            <Button
              onPress={() => this.openRepoUrl(repo.url)}
              title='View Repo on GitHub'
              color={colors.primary}
            />
          </ScrollView>
        </View>
      </View>
    )
  }
}

RepoDetailView.navigationOptions = ({navigation}) => ({
  headerTitle: `${navigation.state.params.repo.name}`
})

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'white'
  },
  hero: {
    height: 200,
    backgroundColor: 'lightgrey',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    padding: 10
  },
  avatar: {
    width: 50,
    height: 50,
    borderRadius: 25,
    marginTop: 20,
    marginBottom: 10
  },
  badgeRow: {
    flex: 1,
    flexDirection: 'row',
    marginTop: 10
  },
  body: {
    flex: 1,
    margin: 20
  },
  scrollView: {
    flex: 1
  }
})
