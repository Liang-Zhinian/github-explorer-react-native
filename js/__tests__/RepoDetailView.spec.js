import 'react-native'
import React from 'react'
import RepoDetailView from '../RepoDetailView'

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer'

/*eslint-disable */
const repo = JSON.parse(
  `
{
  "repo": {
    "id": "MDEwOlJlcG9zaXRvcnkyMTI2MjQ0",
    "name": "bootstrap",
    "description": "The most popular HTML, CSS, and JavaScript framework for developing responsive, mobile first projects on the web.",
    "url": "https://github.com/twbs/bootstrap",
    "homepageUrl": "http://getbootstrap.com",
    "license": "MIT License",
    "owner": {
      "avatarUrl": "https://avatars3.githubusercontent.com/u/2918581?v=3",
      "login": "twbs"
    },
    "stargazers": {
      "totalCount": 110759
    },
    "watchers": {
      "totalCount": 6939
    },
    "primaryLanguage": {
      "color": "#f1e05a",
      "name": "JavaScript"
    }
  }
}
`
)
/*eslint-enable */

const fixture = {
  state: {
    params: repo
  }
}

it('RepoDetailView renders correctly', () => {
  const tree = renderer.create(<RepoDetailView navigation={fixture} />).toJSON()
  expect(tree).toMatchSnapshot()
})
