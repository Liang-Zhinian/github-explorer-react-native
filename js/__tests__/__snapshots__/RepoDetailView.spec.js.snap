// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepoDetailView renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "lightgrey",
        "flexDirection": "column",
        "height": 200,
        "justifyContent": "center",
        "padding": 10,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#EC5D57",
          "fontSize": 42,
        }
      }
    >
      bootstrap
    </Text>
    <Image
      source={
        Object {
          "uri": "https://avatars3.githubusercontent.com/u/2918581?v=3",
        }
      }
      style={
        Object {
          "borderRadius": 25,
          "height": 50,
          "marginBottom": 10,
          "marginTop": 20,
          "width": 50,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      by 
      twbs
    </Text>
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "marginTop": 10,
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "paddingRight": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 16,
              },
              Object {
                "color": "#f1e05a",
                "paddingRight": 2,
              },
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
        >
          
        </Text>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "dimgrey",
              "fontSize": 12,
            }
          }
        >
          JavaScript
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "paddingRight": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 16,
              },
              Object {
                "color": "grey",
                "paddingRight": 2,
              },
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
        >
          
        </Text>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "dimgrey",
              "fontSize": 12,
            }
          }
        >
          110759
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "paddingRight": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 16,
              },
              Object {
                "color": "grey",
                "paddingRight": 2,
              },
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
        >
          
        </Text>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "dimgrey",
              "fontSize": 12,
            }
          }
        >
          6939
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "margin": 20,
      }
    }
  >
    <RCTScrollView
      accessibilityLabel="scrollview"
      alwaysBounceHorizontal={undefined}
      alwaysBounceVertical={true}
      onContentSizeChange={null}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={undefined}
      onResponderTerminationRequest={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      onScrollShouldSetResponder={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      scrollEventThrottle={undefined}
      sendMomentumEvents={false}
      style={
        Array [
          Object {
            "flexDirection": "column",
            "flexGrow": 1,
            "flexShrink": 1,
            "overflow": "scroll",
          },
          Object {
            "flex": 1,
          },
        ]
      }
      testID="scrollview"
    >
      <RCTScrollContentView
        collapsable={false}
        removeClippedSubviews={undefined}
        style={
          Array [
            undefined,
            undefined,
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "marginBottom": 20,
            }
          }
        >
          The most popular HTML, CSS, and JavaScript framework for developing responsive, mobile first projects on the web.
        </Text>
        <View
          accessibilityComponentType="button"
          accessibilityLabel={undefined}
          accessibilityTraits={
            Array [
              "button",
            ]
          }
          accessible={true}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <View
            style={
              Array [
                Object {},
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "#0C42FD",
                    "fontSize": 18,
                    "padding": 8,
                    "textAlign": "center",
                  },
                  Object {
                    "color": "#EC5D57",
                  },
                ]
              }
            >
              View Repo on GitHub
            </Text>
          </View>
        </View>
      </RCTScrollContentView>
    </RCTScrollView>
  </View>
</View>
`;
